### PR TITLE
Misc failsafe changes

### PR DESCRIPTION
**Pull request by dktcoding**
This are another unsorted changes, _most_ of them have a reason to be:

- `Matrix#equals()` wasn't checking for type or `null`
- Since we store Matrix instances in a `Map<String, Matrix>` when serializing multiple matrices, we need a proper implementation of `Matrix#hashCode()` to avoid problems when deserializing in different `libai` instances (yes, the key of the `Map` is a `String` and not a `Matrix`, but not all maps are deserialized the same way).
- Lot's of resources weren't being closed (the `MySQLDataSet` was mostly left alone, but no statement or `ResultSet` is being properly closed, I didn't touch it, because it uses the `ResultSet` itself inside the iterator and didn't have enough time to properly test it)
- I will really start thinking about a `NNIO` class
- Add `serialVersionUID` to all serializable classes (I files serialized with `libai` compiled with OpenJDK don't work in versions compiled with OracleJDK because they lack `serialVersionUID`... I tried... *sigh*)
- About the `@Override` annotations, since I didn't write `libai`, it's fairly difficult to follow inherited methods (I'm sure that I'm not the only one).

### CODE REVIEW

File: src/libai/classifiers/bayes/NaiveBayes.java [hunk 1]
```diff
@@ -113,9 +113,9 @@ private void precalculate(DataSet ds) {
 					double mean = acum.first;
 					double length = (double) ((Integer) data[outputIndex]);
 
-					sd = (sd - ((mean * mean) / (double) length));
-					acum.second = sd / (double) (length - 1);
-					acum.first = mean / (double) length;
+					sd = (sd - ((mean * mean) / length));
```
Comment (kronenthaler) [inline src/libai/classifiers/bayes/NaiveBayes.java]: Please roll back this change. I know it seems silly but there was something not working as expected without the castings. Until we have proper unit test for this classes, i rather keep the changes to the minimum.

Comment (dktcoding) [inline src/libai/classifiers/bayes/NaiveBayes.java]: Ok! reverted.

**Comment by dktcoding** [comment]
About the `@Overrides`, I honestly started finding them useful when reading what other people wrote.

Regarding `ResultSet`s, closing them becomes important with most cached drivers (like the latest driver for MariaDB), and when using connections pools where `Connection#close()` actually releases the lease on the connection but doesn't actually close it.

**Comment by kronenthaler** [comment]
I will check the resultset issues when i'm checking the DataSet refactors.

**Pull request merged**